File: bench/cat.lua
-- BOT-GENERATED

local chunk_size = 64 * 1024

local function copy(input)
  while true do
    local chunk, err = input:read(chunk_size)
    if chunk == nil then
      if err ~= nil then error(err, 0) end
      return
    end
    assert(io.stdout:write(chunk))
  end
end

if #arg == 0 then
  copy(io.stdin)
else
  for index = 1, #arg do
    local name = arg[index]
    if name == "-" then
      copy(io.stdin)
    else
      local input = assert(io.open(name, "rb"))
      copy(input)
      assert(input:close())
    end
  end
end

assert(io.stdout:flush())

File: bench/cat.fs
\ BOT-GENERATED

64 1024 * constant buffer-size
create copy-buffer buffer-size allot

: copy-input ( fileid -- )
  begin
    dup copy-buffer buffer-size rot read-file throw
    dup
  while
    copy-buffer swap stdout write-file throw
  repeat
  2drop
;

: copy-arg { arg arg-len -- }
  arg arg-len s" -" compare 0= if
    stdin copy-input
    exit
  then

  arg arg-len r/o bin open-file throw { input }
  input copy-input
  input close-file throw
;

variable copied

: main ( -- )
  false copied !

  begin
    next-arg 2dup or
  while
    copy-arg
    true copied !
  repeat
  2drop

  copied @ 0= if
    stdin copy-input
  then
  stdout flush-file throw
;

main bye
